### ISSUE TITLE
correct approach for scipy support

### ISSUE THREAD

**Issue by kaoh** [fredrik-johansson/mpmath#393]
I have patched for my own needs scipy to work together with mpmath. I need arbitrary precision math.

Currently I have only quick fixed the minimize BFGS function. scipy is using the `isinf` function which does not know anything about mpmath and fails. I just catched the TypeError for now:

```
        try:
            if isinf(rhok):  # this is patch for numpy
                rhok = 1000.0
                if disp:
                    print("Divide-by-zero encountered: rhok assumed large")
        except TypeError:
            pass
```
See https://github.com/kaoh/scipy/blob/master/scipy/optimize/optimize.py#L1019

What would be a proper way to support this? So far the only way I know would be introspection.

1. Checking if the type the number is not `mpc`, `mpf` (anything else)
2. In the case of mpc or mpf use `isfinite()` of mpmath to check the condition. But this requires a more complex introspection call

**Comment by warnerjon12** [comment]
Yeah, if scipy does not have a similar convert function that generally will try and convert other types to scipy compatible types, then it really doesn't make sense to put arbitrary conversion code anywhere else. I suppose you could add such a function, but that would potentially create a lot of calls to that function. I don't know on this one, I am not familiar enough with scipy.
